### ISSUE TITLE
Wrap vendor specific RNG implementations

### ISSUE THREAD

**Issue by bernhardmgruber** [alpaka-group/alpaka#1798]
We received feedback that it would be great if alpaka could wrap vendor specific RNG implementations like curand, rocrand, or MKL, under a common API.

**Comment by j-stephan** [comment]
This is a duplicate of #1794.

**Comment by SimeonEhrig** [comment]
I would not say duplicate. It extends issue #1794.

**Comment by j-stephan** [comment]
Okay, convinced. But I think we had a similar issue in the past, let's see if I can find it.

**Comment by bernhardmgruber** [comment]
> This is a duplicate of #1794.

It isn't. The referenced issue is about linear algebra libraries and FFT. And even if it mentions `etc.` in the title, I would like to emphasis the request for the RNG libraries explicitly.

**Comment by j-stephan** [comment]
I think I remember now. Didn't we have a discussion in the past to take out the RNG parts of alpaka and move them to a separate library? The motivation behind this was that RNG isn't really a part of the kernel language and thus doesn't fit alpaka's scope.

**Comment by sliwowitz** [comment]
I would agree this is a separate issue mostly because already wrap *some* of the vendor RNGs. There's Xorwow for CUDA/HIP and there are several generators from STL. Originally I thought about making a generic wrapper over all `std` (or `std`-compatible) generators, since it seemed esay due to their unified API, but we decided against it (I don't recall the specifics, but I think the general feeling was those would be unnecessary). With the other vendor-specific generators, the situation is a bit more difficult since, e.g. in cuRand, some generators' `curand_init` take different parameters, etc.

**Comment by sliwowitz** [comment]
> Didn't we have a discussion in the past to take out the RNG parts of alpaka and move them to a separate library?

Yes. Initially, I was implementing Philox in a separate repo as a library. At some point, we decided to include it in `alpaka` though I don't remember what was our reasoning at that time.

**Comment by bernhardmgruber** [comment]
> Initially, I was implementing Philox in a separate repo as a library. At some point, we decided to include it in alpaka though I don't remember what was our reasoning at that time.

I remember arguing for that to make the user experience easier, since no separate library installation would be necessary.

**Comment by bernhardmgruber** [comment]
> Originally I thought about making a generic wrapper over all std (or std-compatible) generators, since it seemed esay due to their unified API, but we decided against it (I don't recall the specifics, but I think the general feeling was those would be unnecessary).

Ok. Well, that was requested now, including the vendor specific APIs :)

>  With the other vendor-specific generators, the situation is a bit more difficult since, e.g. in cuRand, some generators' curand_init take different parameters, etc.

Sure, but what if you really just need a bunch of random numbers with the fastest solution on each backend? And you would not care about determinism across backends. I guess we could choose some reasonable defaults.

**Comment by sliwowitz** [comment]
> Sure, but what if you really just need a bunch of random numbers with the fastest solution on each backend? And you would not care about determinism across backends. I guess we could choose some reasonable defaults.

I don't think this is the actual use case. The Philox should be reasonably fast, also we have the HIP/CUDA Xorwow there, and if one of the other vendor-specific specific RNGs is just *faster* without any other special requirement, then it would make sense to replace Xorwow wit that other one, and be done with it. As I understood it, they wanted to have a wrapper around the vendor-specific parts in order to compare between a native CUDA/whatever vs. alpaka implementation of the whole code in where the RNGs give out exact same numbers (and have the same performance). The "give me a million numbers; quick." part of the discussion was imho more about a higher-level API which we might rather implement in `vikunja`

**Comment by psychocoderHPC** [comment]
I do not think that the user always is looking to fast vendor generators. Sometimes special properties are important e.g. how stable the generator is, how long the unique random sequence is, ...
Besides that someone must wrap different generators into our RNG generators as we already did for CUDA Xorwow I see no reason to not implement other methods.
RNG is nothing that can be moved to vikunja because the RNG is vendor/backend-specific and IMO we do not want to deal with this in vikunja.

Providing methods to fast fill a buffer with rng number is something that can be part of vikunja but this should use low level API's from alpaka.
